### PR TITLE
Qsort from reference article -- rebased

### PR DESCRIPTION
**Pull request by jucor**
Rebase of #147  in a single commit. Conversation started there, continuing here.

### CODE REVIEW

File: lib/TH/generic/THTensorMath.c [hunk 1]
```diff
@@ -1057,64 +1057,135 @@ void THTensor_(reshape)(THTensor *r_, THTensor *t, THLongStorage *size)
...
+
+   Note that other partition schemes exist, and are typically presented
+   in textbook, but those are less efficient. See e.g.
+   http://cs.stackexchange.com/questions/11458/quicksort-partitioning-hoare-vs-lomuto
+
+   Julien, November 12th 2013
 */
-#define  MAX_LEVELS  300
+#define MAX_LEVELS  300
+#define M_SMALL 10 /* Limit for small subfiles */
```
Comment (soumith) [inline lib/TH/generic/THTensorMath.c]: this has not been undefined below anywhere. you might have to add a #undef around line 1179

**Comment by soumith** [comment]
one tiny comment. otherwise looks good and ready for merge!

**Comment by jucor** [comment]
Will add -- and will also add the descending order, now (I had left the old descending for comparison purposes). Do you want all of this in a single commit? In which case, if I understand github correctly, that means opening a new pull-request with a new branch, since rebasing a publicly-pushed commit is prone to screw things up. Do you confirm both points?

**Comment by jucor** [comment]
PS: good catch for the undef, thanks :)

**Comment by soumith** [comment]
git push --force makes it possible to do it in the same PR

**Comment by jucor** [comment]
Here are two extra commits to make your review easier, then I will rebase.

Regarding the `push --force`,  I know that it works, but [Github's help warns of dire consequences](https://help.github.com/articles/using-git-rebase/#pushing-rebased-code-to-github) if other people have checked out the branch in the meantine:

> Force pushing has serious implications, especially if your repository is being accessed by multiple people. At the least, your commits in pull requests might be out of order. For more information, check out ["Why are my commits in the wrong order?"](https://help.github.com/articles/why-are-my-commits-in-the-wrong-order/)

and this other [Githubs' help page on rebase](https://help.github.com/articles/about-git-rebase/) states
> Warning: Because changing your commit history can make things difficult for everyone else using the repository, it's considered bad practice to rebase commits when you've already pushed to a repository.

As I understand it, with the new contribution policy, either:
* either you encourage rebase then `push --force` in PR, at the risk of having messed up histories,
* or you ask every contributor to create a new PR for each rebase occurring in the discussion, which might be a slight hassle.

**Suggestion**: A potentially better middle ground policy could be to keep multiple commits (usual way) *for the time of the discussion until the PR is ready to be merged*,  ask for a rebase into a new branch going into a new PR only once it is ready to merge. This has the side effect of keeping the discussion properly in line with the code changes that they entailed, for easier reference and explanation in the future.

What do you think?

**Comment by soumith** [comment]
push --force is completely okay with PRs because no one uses your interim branch (that is created for the PR). It makes it pretty convenient and seamless, rather than multiple PRs one after another.

your changes look good to me, ready to be rebased.

github's warnings are valid if someone does --force on a master branch that others pull and clone from.

**Comment by nicholas-leonard** [comment]
@jucor @soumith The --force on the PR could indeed cause issues if say, I want to help jucor with his PR by pulling the branch of this repo associated to this PR into a branch of my own repo (or a fork of his). If I make changes and then submit a PR to his branch after he --force rebased it, then yeah, issues. But that doesn't really happen here. And anyone that wants to help out should say so in order to prevent these kinds of misunderstandings...

**Comment by soumith** [comment]
@jucor needs a re-rebase. ping ping.

**Comment by jucor** [comment]
Woops, sorry @soumith. Looks like my antispam ate the github notifications. Thanks for the clarifications on --force, will squash and push in a minute.

**Comment by jucor** [comment]
Done, ready to merge :)

**Comment by soumith** [comment]
@jucor for the win!

**Comment by jucor** [comment]
Hahaha :) Thanks @soumith for your reactivity!

**Pull request merged**